File: common/src/main/java/dev/obscuria/fragmentum/v2/core/common/signal/_Signal3.java
package dev.obscuria.fragmentum.v2.core.common.signal;

import dev.obscuria.fragmentum.v2.api.common.signal.Signal3;

public final class _Signal3<P1, P2, P3> extends _Signal<Signal3.Listener<P1, P2, P3>> implements Signal3<P1, P2, P3> {

    @Override public void emit(P1 p1, P2 p2, P3 p3) {
        emitInternal(it -> it.consume(p1, p2, p3));
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal3.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.tools.signal

class Signal3<P1, P2, P3> : Signal<Signal3.Listener<P1, P2, P3>>() {

    fun emit(p1: P1, p2: P2, p3: P3): Unit = error("API stub")

    fun interface Listener<P1, P2, P3> {

        fun consume(p1: P1, p2: P2, p3: P3)
    }
}